teachertool/src/services/makecodeEditorService.ts

```diff
@@ -0,0 +1,52 @@
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+let makecodeEditorRef: HTMLIFrameElement | undefined;
+const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+let nextId: number = 0;
+let pendingMessages: {[index: string]: PendingMessage} = {};
+
+const onMessageReceived = (event: MessageEvent) => {
+    const data = event.data as pxt.editor.EditorMessageRequest;
+    if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+        const pending = pendingMessages[data.id];
+        pending.handler(data);
+        delete pendingMessages[data.id];
+        return;
+    }
+}
+
+export const sendMessageAsync = (message?: any) => {
+    return new Promise(resolve => {
+        const sendMessageCore = (message: any) => {
+            message.response = true;
+            message.id = nextId++ + "";
+            pendingMessages[message.id] = {
+                original: message,
+                handler: resolve
+            };
+            makecodeEditorRef!.contentWindow!.postMessage(message, "*");
+        }
+
+        if (makecodeEditorRef) {
+            messageQueue.push(message);
+            while (messageQueue.length) {
+                sendMessageCore(messageQueue.shift());
+            }
+            if (message) sendMessageCore(message);
+        }
+    });
+}
+
+export const getIframeRef = (ref: HTMLIFrameElement) => {
+    makecodeEditorRef = ref;
```

Comment:
When the iframe is set, if there are pending messages in the queue, will they be sent to it?

Reply:
So the suggestion here is that if `pendingMessages` is not empty, flush the message queue to the frame when we get a new ref?

Reply:
Yup (unless I misunderstood Eric's original comment)! You can see skillmap doing the same thing here, after it sets the workspace to ready: https://github.com/microsoft/pxt/blob/master/skillmap/src/components/makecodeFrame.tsx#L180

Reply:
I think we might need a rename of `pendingMessages` because I was a bit confused by your question, but I think that you're meaning that if there were any messages sent to the iframe before the iframe was loaded, we would flush those messages once the iframe's loaded. Is that right? Thomas' reply below and some experimenting gave me a bit more  context on this. The "pending messages" to the iframe will exist in the `messageQueue`, or at least should be based on the conversation we have/had in a thread below. Pending messages definitely can get sent to the iframe once it is loaded.  I think this ties into the stream of what should we do when the iframe is ready.

Reply:
Assuming the intention is to match skillmap's behavior, there are two distinct queues here:

`pendingMessages` is the list of messages that _have already been sent_ to the editor and we're waiting for a response
`messageQueue` is the list of messages that _have not yet been sent_ and are waiting for the editor to be ready to receive messages before we send them.

I think we want to clear the `messageQueue` here but not `pendingMessages`. The terminology is a bit confusing, yeah. Perhaps we can think of better names...